pkg/cmd/cli/cmd/debug.go

```diff
@@ -0,0 +1,565 @@
+	AsRoot          bool
+	KeepLabels      bool
+	KeepAnnotations bool
+	KeepLiveness    bool
+	KeepReadiness   bool
+	OneContainer    bool
+	NodeName        string
+	AddEnv          []kapi.EnvVar
+	RemoveEnv       []string
+}
+
+const (
+	debugLong = `
+Launch a command shell to debug a running application
+
+When debugging images and setup problems, it's useful to get an exact copy of a running
+pod configuration and troubleshoot with a shell. Since a pod that is failing may not be
+started and not accessible to 'rsh' or 'exec', the 'debug' command makes it easy to
+create a carbon copy of that setup.
+
+The default mode is to start a shell inside of the first container of the referenced pod,
+replication controller, or deployment config. The started pod will be a copy of your
+source pod, with labels stripped, the command changed to '/bin/sh', and readiness and
+liveness checks disabled. If you just want to run a command, pass -I to disable the
+shell and add '--' and a command to run. Other flags are supported for altering the
+container or pod in common ways.
+
+The debug pod is deleted when the the remote command completes or the user interrupts
+the shell.`
+
+	debugExample = `
+  # Debug a currently running deployment
+  $ %[1]s dc/test
+
+  # Debug a specific failing container by running the env command in the 'second' container
+  $ %[1]s dc/test -c second -- /bin/env
+
+  # See the pod that would be created to debug
+  $ %[1]s dc/test -o yaml`
+
+	debugPodLabelName = "debug.openshift.io/name"
+
+	debugPodAnnotationSourceContainer = "debug.openshift.io/source-container"
+	debugPodAnnotationSourceResource  = "debug.openshift.io/source-resource"
+)
+
+// NewCmdDebug creates a command for debugging pods.
+func NewCmdDebug(fullName string, f *clientcmd.Factory, in io.Reader, out, errout io.Writer) *cobra.Command {
+	options := &DebugOptions{
+		Timeout: 30 * time.Second,
+		Command: []string{"/bin/sh"},
+		Attach: kcmd.AttachOptions{
+			In:    in,
+			Out:   out,
+			Err:   errout,
+			TTY:   true,
+			Stdin: true,
+
+			Attach: &kcmd.DefaultRemoteAttach{},
+		},
+		LogsForObject: f.LogsForObject,
+	}
+
+	cmd := &cobra.Command{
+		Use:     "debug RESOURCE/NAME [ENV1=VAL1 ...] [-c CONTAINER] [-- COMMAND]",
+		Short:   "Launch a new instance of a pod for debugging",
+		Long:    debugLong,
+		Example: fmt.Sprintf(debugExample, fmt.Sprintf("%s debug", fullName)),
+		Run: func(cmd *cobra.Command, args []string) {
+			kcmdutil.CheckErr(options.Complete(cmd, f, args, in, out, errout))
+			kcmdutil.CheckErr(options.Validate())
+			kcmdutil.CheckErr(options.Debug())
+		},
+	}
+
+	// TODO: when T is deprecated use the printer, but keep these hidden
+	cmd.Flags().StringP("output", "o", "", "Output format. One of: json|yaml|wide|name|go-template=...|go-template-file=...|jsonpath=...|jsonpath-file=... See golang template [http://golang.org/pkg/text/template/#pkg-overview] and jsonpath template [http://releases.k8s.io/HEAD/docs/user-guide/jsonpath.md].")
```

Comment:
It's because of -t.  I don't need to keep the deprecated flag for template
since this is net new.  We have no better way to express this.

On Fri, Feb 19, 2016 at 3:53 PM, David Eads <notifications@github.com>
wrote:

> In pkg/cmd/cli/cmd/debug.go
> <https://github.com/openshift/origin/pull/7294#discussion_r53516645>:
>
> > +	}
> > +
> > +	cmd := &cobra.Command{
> > +		Use:     "debug RESOURCE/NAME [ENV1=VAL1 ...] [-c CONTAINER] [-- COMMAND]",
> > +		Short:   "Launch a new instance of a pod for debugging",
> > +		Long:    debugLong,
> > +		Example: fmt.Sprintf(debugExample, fmt.Sprintf("%s debug", fullName)),
> > +		Run: func(cmd *cobra.Command, args []string) {
> > +			kcmdutil.CheckErr(options.Complete(cmd, f, args, in, out, errout))
> > +			kcmdutil.CheckErr(options.Validate())
> > +			kcmdutil.CheckErr(options.Debug())
> > +		},
> > +	}
> > +
> > +	// TODO: when T is deprecated use the printer, but keep these hidden
> > +	cmd.Flags().StringP("output", "o", "", "Output format. One of: json|yaml|wide|name|go-template=...|go-template-file=...|jsonpath=...|jsonpath-file=... See golang template [http://golang.org/pkg/text/template/#pkg-overview] and jsonpath template [http://releases.k8s.io/HEAD/docs/user-guide/jsonpath.md].")
>
> many of these look like cruft
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/7294/files#r53516645>.
>